File: python/0530二叉搜索树的最小绝对差.py
from binarytree import TreeNode,level_create

def getMinimumDifference(root: TreeNode) -> int:
    res=float("inf")
    pre=float("inf")
    def _traversal(root:TreeNode):
        nonlocal res
        nonlocal pre

        if(not root):
            return

        if(root.left):
            _traversal(root.left)
        
        res=min(abs(pre-root.val),res)
        pre=root.val

        if(root.right):
            _traversal(root.right)

    _traversal(root)

    return res


print(getMinimumDifference(level_create([236,104,701,None,227,None,911])))

File: cpp/0530二叉搜索树的最小绝对差.cpp
#include"mybitree.h"
#include<stack>
#include<vector>
#include<algorithm>

int getMinimumDifference(TreeNode *root)
{
    stack<TreeNode *> st;
    st.push(root);
    int res = 0x7fffffff;
    int pre = 0x7fffffff;
    vector<TreeNode *> memo;
    while (!st.empty())
    {
        TreeNode *crt = st.top();

        if (crt->left && find(memo.begin(), memo.end(), crt->left) == memo.end())
        {
            st.push(crt->left);
            continue;
        }
        else
        {
            st.pop();
            res = min(abs(pre - crt->val), res);
            pre = crt->val;
            memo.push_back(crt);
        }
        if(crt->right)
            st.push(crt->right);
    }
    return res;
}